### ISSUE TITLE
unnecessary trainign ops in final model after freeze graph

### ISSUE THREAD

**Issue by yasingh2007** [yinguobing/head-pose-estimation#31]
@yinguobing I listed the ops from your prebuilt " frozen_inference_graph.pb" and freeze_graph created from your saved_model "head-pose-estimation/assets/pose_model/".

How did you do the optimization/pruning during freezing. what option did you use with tools.freeze_graph to remove unnecessary training ops?

Below are the difference between your saved model and frozen_inference_graph.pb

saved_model                                frozen_inference_graph.pb

1."image_tensor"   (?,?,?,?3) ---> input_image_tensor (128,128,3)
2. All "map" nodes replaced by --> Reshape,input_to_float
3. layer6/final_dense removed
4."layer" name removed from ops, how and but why?
5. conv2d names --> postfix with "conv2d_1" till "conv2d_8".

What are the steps to do optimization and pruning to get "frozen_inference_graph.pb" like freezed graph?

python3 -m tensorflow.python.tools.freeze_graph --input_saved_model_dir ../head-pose-estimation/assets/pose_model/ --output_node_names layer6/final_dense     --output_graph frozen_graph.pb
python3 demo.py --model frozen_graph.pb --list_ops true 
butterfly/image_tensor
butterfly/map/Shape
butterfly/map/strided_slice/stack
butterfly/map/strided_slice/stack_1
butterfly/map/strided_slice/stack_2
butterfly/map/strided_slice
butterfly/map/TensorArray
butterfly/map/TensorArrayUnstack/Shape
butterfly/map/TensorArrayUnstack/strided_slice/stack
butterfly/map/TensorArrayUnstack/strided_slice/stack_1
butterfly/map/TensorArrayUnstack/strided_slice/stack_2
butterfly/map/TensorArrayUnstack/strided_slice
butterfly/map/TensorArrayUnstack/range/start
butterfly/map/TensorArrayUnstack/range/delta
butterfly/map/TensorArrayUnstack/range
butterfly/map/TensorArrayUnstack/TensorArrayScatter/TensorArrayScatterV3
butterfly/map/Const
butterfly/map/TensorArray_1
butterfly/map/while/iteration_counter
butterfly/map/while/Enter
butterfly/map/while/Enter_1
butterfly/map/while/Enter_2
butterfly/map/while/Merge
butterfly/map/while/Merge_1
butterfly/map/while/Merge_2
butterfly/map/while/Less/Enter
butterfly/map/while/Less
butterfly/map/while/Less_1
butterfly/map/while/LogicalAnd
butterfly/map/while/LoopCond
butterfly/map/while/Switch
butterfly/map/while/Switch_1
butterfly/map/while/Switch_2
butterfly/map/while/Identity
butterfly/map/while/Identity_1
butterfly/map/while/Identity_2
butterfly/map/while/add/y
butterfly/map/while/add
butterfly/map/while/TensorArrayReadV3/Enter
butterfly/map/while/TensorArrayReadV3/Enter_1
butterfly/map/while/TensorArrayReadV3
butterfly/map/while/resize/ExpandDims/dim
butterfly/map/while/resize/ExpandDims
butterfly/map/while/resize/size
butterfly/map/while/resize/ResizeBilinear
butterfly/map/while/resize/Squeeze
butterfly/map/while/TensorArrayWrite/TensorArrayWriteV3/Enter
butterfly/map/while/TensorArrayWrite/TensorArrayWriteV3
butterfly/map/while/add_1/y
butterfly/map/while/add_1
butterfly/map/while/NextIteration
butterfly/map/while/NextIteration_1
butterfly/map/while/NextIteration_2
butterfly/map/while/Exit_2
butterfly/map/TensorArrayStack/TensorArraySizeV3
butterfly/map/TensorArrayStack/range/start
butterfly/map/TensorArrayStack/range/delta
butterfly/map/TensorArrayStack/range
butterfly/map/TensorArrayStack/TensorArrayGatherV3
butterfly/layer1/conv2d/kernel
butterfly/layer1/conv2d/kernel/read
butterfly/layer1/conv2d/bias
butterfly/layer1/conv2d/bias/read
butterfly/layer1/conv2d/Conv2D
butterfly/layer1/conv2d/BiasAdd
butterfly/layer1/conv2d/Relu
butterfly/layer1/max_pooling2d/MaxPool
butterfly/layer2/conv2d/kernel
butterfly/layer2/conv2d/kernel/read
butterfly/layer2/conv2d/bias
butterfly/layer2/conv2d/bias/read
butterfly/layer2/conv2d/Conv2D
butterfly/layer2/conv2d/BiasAdd
butterfly/layer2/conv2d/Relu
butterfly/layer2/conv2d_1/kernel
butterfly/layer2/conv2d_1/kernel/read
butterfly/layer2/conv2d_1/bias
butterfly/layer2/conv2d_1/bias/read
butterfly/layer2/conv2d_1/Conv2D
butterfly/layer2/conv2d_1/BiasAdd
butterfly/layer2/conv2d_1/Relu
butterfly/layer2/max_pooling2d/MaxPool
butterfly/layer3/conv2d/kernel
butterfly/layer3/conv2d/kernel/read
butterfly/layer3/conv2d/bias
butterfly/layer3/conv2d/bias/read
butterfly/layer3/conv2d/Conv2D
butterfly/layer3/conv2d/BiasAdd
butterfly/layer3/conv2d/Relu
butterfly/layer3/conv2d_1/kernel
butterfly/layer3/conv2d_1/kernel/read
butterfly/layer3/conv2d_1/bias
butterfly/layer3/conv2d_1/bias/read
butterfly/layer3/conv2d_1/Conv2D
butterfly/layer3/conv2d_1/BiasAdd
butterfly/layer3/conv2d_1/Relu
butterfly/layer3/max_pooling2d/MaxPool
butterfly/layer4/conv2d/kernel
butterfly/layer4/conv2d/kernel/read
butterfly/layer4/conv2d/bias
butterfly/layer4/conv2d/bias/read
butterfly/layer4/conv2d/Conv2D
butterfly/layer4/conv2d/BiasAdd
butterfly/layer4/conv2d/Relu
butterfly/layer4/conv2d_1/kernel
butterfly/layer4/conv2d_1/kernel/read
butterfly/layer4/conv2d_1/bias
butterfly/layer4/conv2d_1/bias/read
butterfly/layer4/conv2d_1/Conv2D
butterfly/layer4/conv2d_1/BiasAdd
butterfly/layer4/conv2d_1/Relu
butterfly/layer4/max_pooling2d/MaxPool
butterfly/layer5/conv2d/kernel
butterfly/layer5/conv2d/kernel/read
butterfly/layer5/conv2d/bias
butterfly/layer5/conv2d/bias/read
butterfly/layer5/conv2d/Conv2D
butterfly/layer5/conv2d/BiasAdd
butterfly/layer5/conv2d/Relu
butterfly/layer6/flatten/Shape
butterfly/layer6/flatten/strided_slice/stack
butterfly/layer6/flatten/strided_slice/stack_1
butterfly/layer6/flatten/strided_slice/stack_2
butterfly/layer6/flatten/strided_slice
butterfly/layer6/flatten/Reshape/shape/1
butterfly/layer6/flatten/Reshape/shape
butterfly/layer6/flatten/Reshape
butterfly/layer6/dense/kernel
butterfly/layer6/dense/kernel/read
butterfly/layer6/dense/bias
butterfly/layer6/dense/bias/read
butterfly/layer6/dense/MatMul
butterfly/layer6/dense/BiasAdd
butterfly/layer6/dense/Relu
butterfly/layer6/logits/kernel
butterfly/layer6/logits/kernel/read
butterfly/layer6/logits/bias
butterfly/layer6/logits/bias/read
butterfly/layer6/logits/MatMul
butterfly/layer6/logits/BiasAdd
butterfly/layer6/final_dense

butterfly$ python3 demo.py --model frozen_inference_graph.pb --list_ops true
butterfly/input_image_tensor
butterfly/Reshape/shape
butterfly/Reshape
butterfly/input_to_float
butterfly/conv2d/kernel
butterfly/conv2d/kernel/read
butterfly/conv2d/bias
butterfly/conv2d/bias/read
butterfly/conv2d/Conv2D
butterfly/conv2d/BiasAdd
butterfly/conv2d/Relu
butterfly/max_pooling2d/MaxPool
butterfly/conv2d_1/kernel
butterfly/conv2d_1/kernel/read
butterfly/conv2d_1/bias
butterfly/conv2d_1/bias/read
butterfly/conv2d_2/Conv2D
butterfly/conv2d_2/BiasAdd
butterfly/conv2d_2/Relu
butterfly/conv2d_2/kernel
butterfly/conv2d_2/kernel/read
butterfly/conv2d_2/bias
butterfly/conv2d_2/bias/read
butterfly/conv2d_3/Conv2D
butterfly/conv2d_3/BiasAdd
butterfly/conv2d_3/Relu
butterfly/max_pooling2d_2/MaxPool
butterfly/conv2d_3/kernel
butterfly/conv2d_3/kernel/read
butterfly/conv2d_3/bias
butterfly/conv2d_3/bias/read
butterfly/conv2d_4/Conv2D
butterfly/conv2d_4/BiasAdd
butterfly/conv2d_4/Relu
butterfly/conv2d_4/kernel
butterfly/conv2d_4/kernel/read
butterfly/conv2d_4/bias
butterfly/conv2d_4/bias/read
butterfly/conv2d_5/Conv2D
butterfly/conv2d_5/BiasAdd
butterfly/conv2d_5/Relu
butterfly/max_pooling2d_3/MaxPool
butterfly/conv2d_5/kernel
butterfly/conv2d_5/kernel/read
butterfly/conv2d_5/bias
butterfly/conv2d_5/bias/read
butterfly/conv2d_6/Conv2D
butterfly/conv2d_6/BiasAdd
butterfly/conv2d_6/Relu
butterfly/conv2d_6/kernel
butterfly/conv2d_6/kernel/read
butterfly/conv2d_6/bias
butterfly/conv2d_6/bias/read
butterfly/conv2d_7/Conv2D
butterfly/conv2d_7/BiasAdd
butterfly/conv2d_7/Relu
butterfly/max_pooling2d_4/MaxPool
butterfly/conv2d_7/kernel
butterfly/conv2d_7/kernel/read
butterfly/conv2d_7/bias
butterfly/conv2d_7/bias/read
butterfly/conv2d_8/Conv2D
butterfly/conv2d_8/BiasAdd
butterfly/conv2d_8/Relu
butterfly/flatten/Shape
butterfly/flatten/strided_slice/stack
butterfly/flatten/strided_slice/stack_1
butterfly/flatten/strided_slice/stack_2
butterfly/flatten/strided_slice
butterfly/flatten/Reshape/shape/1
butterfly/flatten/Reshape/shape
butterfly/flatten/Reshape
butterfly/dense/kernel
butterfly/dense/kernel/read
butterfly/dense/bias
butterfly/dense/bias/read
butterfly/dense/MatMul
butterfly/dense/BiasAdd
butterfly/dense/Relu
butterfly/logits/kernel
butterfly/logits/kernel/read
butterfly/logits/bias
butterfly/logits/bias/read
butterfly/logits/MatMul
butterfly/logits/BiasAdd

**Comment by yasingh2007** [comment]
@yinguobing ????

**Comment by yasingh2007** [comment]
@yinguobing What is your method to create "head-pose-estimation/assets/frozen_inference_graph.pb"? I used "python3 -m tensorflow.python.tools.freeze_graph --input_saved_model_dir ../head-pose-estimation/assets/pose_model/ --output_node_names layer6/final_dense --output_graph frozen_graph.pb". But it produce graph different then your frozen_inference_graph.p (have "map/*" training nodes.

**Comment by yinguobing** [comment]
It's not me, I didn't do anything special.

Follow these [instructions](https://github.com/yinguobing/cnn-facial-landmark) then you should obtain the same result.

Also make sure the version of TensorFlow is 1.15.

**Comment by yasingh2007** [comment]
@yinguobing Thanks, your comment is really helpful.it will now save me some time by not editing the graph through TransformGrahp(),remove_training_nodes(), optimize_for_inferencing() (I tried to do that to remove map/* nodes without success). 

1)I followed the same commands from the link above. however I find that --eval_only option is not implemented in current landmark.py and I have build without it. I am using same version of TF. 
(base) ubuntu:~$ pip list | grep tensor
tensorboard 1.15.0
**tensorflow 1.15.2**
tensorflow-datasets 3.1.0
tensorflow-estimator 1.15.1
tensorflow-gpu 1.15.2
tensorflow-metadata 0.22.0
tensorrt 7.0.0.11

2. Your frozen_inference_graph.pb has  first layer name as "input_image_tensor" and final layer "logits/BiasAdd". but current landmark.py has "image_tensor" and "logits/final_dense".
Is frozen_inference_graph.pb is made from older version of landmark.py, model.py which version, any date from git history?

3) I believe you have not used optimize_for_infernecing() and only command tool freeze_graph.py to freeze the graph?
